File: flink-learning-example/src/main/java/com/hochoy/flink/socket/SocketWindowWordCount.java
package com.hochoy.flink.socket;

import org.apache.flink.api.common.functions.FlatMapFunction;
import org.apache.flink.api.common.functions.ReduceFunction;
import org.apache.flink.api.java.tuple.Tuple2;
import org.apache.flink.api.java.utils.ParameterTool;
import org.apache.flink.streaming.api.datastream.DataStream;
import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
import org.apache.flink.streaming.api.windowing.time.Time;
import org.apache.flink.util.Collector;

import java.util.Arrays;

public class SocketWindowWordCount {

    // nc -l -p 9999
    // --hostname localhost --port 9999
    public static void main(String[] args) {
        final String hostname;
        final int port;
        try {
            final ParameterTool params = ParameterTool.fromArgs(args);
            hostname = params.has("hostname") ? params.get("hostname") : "localhost";
            port = params.getInt("port");
        } catch (Exception e) {
            System.err.println("No port specified. Please run 'SocketWindowWordCount " +
                    "--hostname <hostname> --port <port>', where hostname (localhost by default) " +
                    "and port is the address of the text server");
            System.err.println("To start a simple text server, run 'netcat -l <port>' and " +
                    "type the input text into the command line");
            return;
        }

        final StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();

        DataStream<String> text = env.socketTextStream(hostname, port, "\n");

        DataStream<Tuple2<String, Long>> wc = text
                .flatMap(new FlatMapFunction<String, Tuple2<String, Long>>() {
                    @Override
                    public void flatMap(String value, Collector<Tuple2<String, Long>> out) throws Exception {
                        Arrays.asList(value.split("\\s")).stream().filter(x->!"".equals(x)).forEach(w -> out.collect(new Tuple2<>(w, 1L)));
                    }
                })
                .keyBy(0)
                .timeWindow(Time.seconds(5))
                .reduce(new ReduceFunction<Tuple2<String, Long>>() {
                    @Override
                    public Tuple2<String, Long> reduce(Tuple2<String, Long> value1, Tuple2<String, Long> value2) throws Exception {
                        return new Tuple2<>(value1.f0, value1.f1 + value2.f1);
                    }
                });

        wc.print().setParallelism(1);

        try {

            env.execute("socket window wordCount");
        }catch (Exception e ){
            e.printStackTrace();
        }

    }

}

File: flink-learning-example/src/main/scala/com/hochoy/flink/streaming/transform/SocketWindowWordCount.scala
package com.hochoy.flink.streaming.transform

import org.apache.commons.lang3.StringUtils
import org.apache.flink.api.java.utils.ParameterTool
import org.apache.flink.streaming.api.scala._
import org.apache.flink.streaming.api.windowing.time.Time

import java.util.Date

object SocketWindowWordCount {

  var host :String = _
  var port :Int = _
  def main(args: Array[String]): Unit = {

    // --host localhost --port 9999

    val tool: ParameterTool = ParameterTool.fromArgs(args)

    host =  if (tool.has("host")) tool.get("host") else  "localhost"
    port = if(tool.has("port")) tool.getInt("port") else 9999

    val env: StreamExecutionEnvironment = StreamExecutionEnvironment.getExecutionEnvironment

    val source: DataStream[String] =   env.socketTextStream(host, port)


//    val windowCounts =
//      source.filter(StringUtils.isNoneBlank(_))
//        .flatMap { w => w.split("\\s") }
//        .filter(StringUtils.isNoneBlank(_))
//      .map { w =>  (w, 1) }
//      .keyBy(0)
//      .timeWindow(Time.seconds(5))
//      .sum(1)
//    windowCounts.print(" >>>>>>> ")

    val split: DataStream[String] = source.filter(e => StringUtils.isNotBlank(e))
        .flatMap(_.split("\\s")).filter(StringUtils.isNotBlank(_))

    val wordAndOne: DataStream[(String, Int)] = split.map((_, 1))

    val summed: DataStream[(String, Int)] =
      wordAndOne
        .keyBy(0)
        .timeWindow(Time.seconds(5))
        .sum(1)
    summed.print("sum>>>   ")
    env.execute()
  }
  case class WordWithCount(word: String, count: Long)
}
